File: backend/src/ai/priority_engine.py
def calculate_priority(data):
    """
    Calculates the priority score and priority level based on various factors.

    Args:
        data (dict): A dictionary containing the priority factors.
            - hazardous (bool): Whether the item is hazardous.
            - weight (float): The weight of the item.
            - daysWaiting (int): The number of days the item has been waiting.
            - locationType (str): The type of location (Hospital, School, Residential, Market).

    Returns:
        dict: A dictionary containing the priority score, priority level, and reasons.
            - score (int): The calculated priority score.
            - priority (str): The priority level (Low, Medium, High, Critical).
            - reasons (list): A list of strings explaining the score calculation.
    """
    score = 0
    reasons = []

    if data["hazardous"]:
        score += 40
        reasons.append("Hazardous item: +40 points")

    weight_score = data["weight"] * 2
    score += weight_score
    reasons.append(f"Weight: {weight_score} points")

    waiting_score = data["daysWaiting"] * 5
    score += waiting_score
    reasons.append(f"Days waiting: {waiting_score} points")

    if data["locationType"] == "Hospital":
        score += 30
        reasons.append("Location type (Hospital): +30 points")
    elif data["locationType"] == "School":
        score += 20
        reasons.append("Location type (School): +20 points")
    elif data["locationType"] == "Residential":
        score += 10
        reasons.append("Location type (Residential): +10 points")
    elif data["locationType"] == "Market":
        score += 15
        reasons.append("Location type (Market): +15 points")

    if score >= 91:
        priority = "Critical"
    elif score >= 61:
        priority = "High"
    elif score >= 31:
        priority = "Medium"
    else:
        priority = "Low"

    return {"score": score, "priority": priority, "reasons": reasons}


if __name__ == "__main__":
    test_data = {
        "hazardous": True,
        "weight": 5.5,
        "daysWaiting": 3,
        "locationType": "Hospital"
    }

    result = calculate_priority(test_data)
    print("\n========== RESULT ==========\n")

    for key, value in result.items():
        if isinstance(value, list):
            print(f"{key}:")
            for item in value:
                print(f"  • {item}")
        else:
            print(f"{key}: {value}")

    print("\n============================")

File: backend/src/ai/priorityEngine.js
/**
 * Calculates the priority score and level for waste collection based on various factors.
 *
 * @param {Object} data - The input data object containing waste collection details.
 * @param {boolean} data.hazardous - Indicates if the waste is hazardous.
 * @param {number} data.weight - The weight of the waste in kilograms.
 * @param {number} data.daysWaiting - The number of days the waste has been waiting for collection.
 * @param {string} data.locationType - The type of location where the waste is located.
 *
 * @returns {Object} An object containing the calculated priority score, priority level, and reasons.
 * @returns {number} score - The calculated priority score.
 * @returns {string} priority - The priority level based on the score (Low, Medium, High, Critical).
 * @returns {string[]} reasons - An array of strings describing the reasons for the priority score.
 */
function calculatePriority(data) {
    // Extract the input data from the data object
    const { hazardous, weight, daysWaiting, locationType } = data;
  
    // Initialize the priority score and reasons array
    let score = 0;
    const reasons = [];
  
    // Check if the waste is hazardous
    if (hazardous) {
      score += 40;
      reasons.push('+40 points for hazardous waste');
    }
  
    // Add points based on the weight of the waste
    score += weight * 2;
    reasons.push(`+${weight * 2} points for ${weight} kg of waste`);
  
    // Add points based on the number of days the waste has been waiting
    score += daysWaiting * 5;
    reasons.push(`+${daysWaiting * 5} points for ${daysWaiting} days waiting`);
  
    // Add points based on the location type
    switch (locationType) {
      case 'Hospital':
        score += 30;
        reasons.push('+30 points for hospital location');
        break;
      case 'School':
        score += 20;
        reasons.push('+20 points for school location');
        break;
      case 'Residential':
        score += 10;
        reasons.push('+10 points for residential location');
        break;
      case 'Market':
        score += 15;
        reasons.push('+15 points for market location');
        break;
    }
  
    // Determine the priority level based on the score
    let priority;
    if (score >= 91) {
      priority = 'Critical';
    } else if (score >= 61) {
      priority = 'High';
    } else if (score >= 31) {
      priority = 'Medium';
    } else {
      priority = 'Low';
    }
  
    // Return the calculated priority score, priority level, and reasons
    return {
      score,
      priority,
      reasons
    };
  }
  
  // Export the calculatePriority function using CommonJS
  module.exports = calculatePriority;